[AIE2P] Legalization and instruction selection support for G_FADD/G_FSUB

katerynamuts opened the pull request.

SagarMaheshwari99 commented on llvm/test/CodeGen/AIE/aie2p/GlobalIsel/inst-select-fadd.mir:
```diff
@@ -0,0 +1,33 @@
+
+
+---
+name:            test_fadd_bfloat16
```
this requires name change.

SagarMaheshwari99 replied:
so do rest of the tests.

SagarMaheshwari99 commented on llvm/test/CodeGen/AIE/GlobalISel/legalize-fsub.mir:
```diff
@@ -24,12 +24,120 @@ body: |
+    ; COMMON-NEXT: [[ASSERT_SEXT:%[0-9]+]]:_(s32) = G_ASSERT_SEXT [[AIE_SEXT_EXTRACT_VECTOR_ELT32]], 16
     ; COMMON-NEXT: $r0 = COPY [[ASSERT_SEXT]](s32)
     ; COMMON-NEXT: PseudoRET implicit $lr, implicit $r0
     %0:_(s32) = COPY $r1
```
Can you add a couple of tests for vectors, for ex. `<16 x s32>` , same for fadd..

katerynamuts committed:
[AIE2P] Instruction selection support for G_FADD/G_FSUB

mludevid committed:
[AIE2P] Legalize G_FADD/G_FSUB

mludevid commented:
I've updated the legalization commit. It now legalizes the non-vector case properly. I've also split up the test for vectors on aie2 and aie2p, because the two architectures perform the `G_FADD` and `G_FSUB` operations on different vector sizes.

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1036,11 +1036,14 @@ bool AIELegalizerHelper::legalizeG_FCMP(
+  if (ST.isAIE2P() && InputRegSize == 1024)
+    return Intrinsic::aie2p_v32accfloat_to_v32bf16;
 
   llvm_unreachable("Called with unknown target triple!");
```
This is not wholly accurate. I would group the isAIE2 and isAIE2P and then have separate unreachables or asserts for InputRegSize.

mludevid replied:
You're right. I have split up the if statements into two levels, first the architecture and then the input register size to give more accurate `llvm_unreachable` messages.

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1165,53 +1170,90 @@ bool AIELegalizerHelper::legalizeG_FADDSUB(LegalizerHelper &Helper,
   const LLT ACC512 = LLT::fixed_vector(8, 64);
+  const LLT ACC1024 = LLT::fixed_vector(32, 32);
+
+  LLT InsertVecLLT;
```
Please const initialize with conditional operator. It will fit on one line.

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1165,53 +1170,90 @@ bool AIELegalizerHelper::legalizeG_FADDSUB(LegalizerHelper &Helper,
+  const LLT V32FP32 = LLT::fixed_vector(32, 32);
+  const LLT V64FP32 = LLT::fixed_vector(64, 32);
   const LLT ACC512 = LLT::fixed_vector(8, 64);
+  const LLT ACC1024 = LLT::fixed_vector(32, 32);
```
V8ACC64 / V32ACC32?

mludevid replied:
Done. I also moved all constant LLT declarations across the entire file into the class declaration to share them across functions.

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1165,53 +1170,90 @@ bool AIELegalizerHelper::legalizeG_FADDSUB(LegalizerHelper &Helper,
+    DstLLT = V32BF16;
   } else {
     FPRes = FPOp;
+    DstLLT = V16BF16;
```
Hmm. In other places the fallback is AIE2. what is it here?

martien-de-jong replied:
I guess this method is begging for virtualization into version specific subclasses.

mludevid replied:
As discussed offline, I totally agree, this is begging for virtualization. But I'm not sure yet whether it would be better to virtualize the entire legalization function or only smaller parts, e.g. scalar to vector, or expand vector to some size, etc. In any case this shouldn't block this PR being merged. For now I have simplified this function as much as possible to minimize the usage of architecture specific code blocks.

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1165,53 +1170,90 @@ bool AIELegalizerHelper::legalizeG_FADDSUB(LegalizerHelper &Helper,
+  }
 
   Register FPOp;
   if (MI.getOpcode() == TargetOpcode::G_FADD)
```
Could we rename the function to legalizeG_FADD_G_FSUB

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1165,53 +1170,90 @@ bool AIELegalizerHelper::legalizeG_FADDSUB(LegalizerHelper &Helper,
-          .addUse(FPRes)
-          .getReg(0);
+  const int VecSize = MRI.getType(FPRes).getSizeInBits();
+
```
drop newline?

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1165,53 +1170,90 @@ bool AIELegalizerHelper::legalizeG_FADDSUB(LegalizerHelper &Helper,
-      MIRBuilder
-          .buildInstr(InsertEltOpc, {V16FP32}, {Src2Vec, NewSrcReg2, IdxReg})
-          .getReg(0);
+  Register NewSrc1 = MIRBuilder
```
nit: NewSrc1 is just as much a register as NewSrcReg1. I wouldn't mind updating NewSrcReg1

martien-de-jong replied:
And then, since all these variables are Registers anyway, drop the Reg?

martien-de-jong commented on llvm/test/CodeGen/AIE/GlobalISel/legalize-fadd.mir:
Is this test redundant now that we have aie2 and aie2p specific versions?

mludevid replied:
This test is for sure not redundant, because it is testing something different. In this file we are testing adding two `bfloat16`s (GPRs). In the aie2 and aie2p specific versions we are testing the addition of two `bfloat16` vectors. What we could do is copy this test case into the two aie2 and aie2p specific test files and check them there instead.

martien-de-jong replied:
That' or have a more telling name, like legalize-scalar-fadd,mir

mludevid replied:
I renamed both test files to `legalize-scalar-fadd.mir` and `legalzie-scalar-fsub.mir`.

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1165,53 +1170,90 @@ bool AIELegalizerHelper::legalizeG_FADDSUB(LegalizerHelper &Helper,
 
-  auto Pad512 =
-      emitPadUndefVector(MRI, MIRBuilder, V16BF16.multiplyElements(2), Conv);
+  Register Reg512;
```
const initialize with conditional operator please. Also, V16BF16.multiplyElements(2) is just V32BF16

mludevid committed:
[NFC][AIEX] Rename G_FADD/G_FSUB legalization func

mludevid committed:
[NFC][AIEX] Decl LLTs in AIELegalizerHelper class

We use most vector LLTs across multiple legalization functions. I moved
their declaration into the AIELegalizerHelper class to be shared across
functions.

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1156,102 +1162,68 @@ bool AIELegalizerHelper::legalizeG_FADD_G_FSUB(LegalizerHelper &Helper,
+  Res = MIRBuilder.buildInstr(ExtractEltOpc, {S32}, {Res, IdxReg}).getReg(0);
+  Res = MIRBuilder.buildAssertInstr(TargetOpcode::G_ASSERT_SEXT, {S32}, Res, 16)
+            .getReg(0);
+  MIRBuilder.buildTrunc(DstReg, Res);
```
Check: Trunc32ToBF16 has left bf16, and the extract has sign-extended it to span a 32 bit reg. we assert that it was properly sign extended, then trunc it back  (to 16 bit unsigned? or is it a noop for type consistency?)

mludevid replied:
We first truncate (convert) the vector of `s32`s (an accumulator) to one of `s16`s (in a vector register) and then extract the first element. Since that returns an `s32` we assert that it was properly sign extended and then truncate it down to `s16` again. But during further legalization this truncation to `s16` is legalized to a copy and after instruction selection the assertion is deleted and only the copy is left.

khallouh commented on llvm/test/CodeGen/AIE/aie2p/GlobalIsel/legalize-fadd.mir:
```diff
@@ -0,0 +1,27 @@
+# RUN: llc -mtriple aie2p -run-pass=legalizer %s -verify-machineinstrs -o - | FileCheck %s
+
+---
+name: test_fadd_vec_2048
```
We didn't support other vector sizes. I suppose that's because we were only interested in S16?
